Volume revert to snapshot

wwba opened the pull request:
volume-revert-to-snapshot

curtbruns committed:
Add deactivate step to extend_lv

Extending a linear LVM volume that has a snapshot requires
that the LV be deactivated explicitly prior to extending.

This adds a deactivate_lv method and calls it prior to
issuing the extend call. If auto_activation_volume_list
is not defined lvm.conf, then volumes will be reactivated
automatically after the extend.  If auto_activation_volume_list
is defined, then volumes that should be automatically
reactivated should be added to the auto_activation_volume_list
or they won't be activated automatically.
NOTE: This doesn't apply to thin provisioned LVs as they don't
require the deactivation step.

DocImpact: When a user extends LVM Volumes with a snapshot, the
volumes will be deactivated.  Their re-activation is automatic,
unless auto_activation_volume_list is defined in lvm.conf.
See lvm.conf for more info.  Thin provisioned LVM Volumes
will not be deactivated as they don't require it.

Co-Authored-By: John Griffith <john.griffith8@gmail.com>
Change-Id: If746625cfe658c3528525dc7f4bf869f1c9704dc
Closes-Bug: #1470558
(cherry picked from commit 314611f79afa6fe79f3d5022c814772048669bb0)

TommyLike committed:
Cinder volume revert to snapshot

Fixes: Feature #10661

This patch implements the spec of reverting volume to
latest snapshot, generic way to revert volume if driver's method
is missing if not supported.
Related tempest and client patches:

[1] https://review.openstack.org/#/c/463906/
[2] https://review.openstack.org/#/c/464903/

APIImpact
DocImpact
Partial-Implements: blueprint revert-volume-to-snapshot

Change-Id: Ib20d749c2118c350b5fa0361ed1811296d518a17
(cherry picked from commit 8fba9a90807714f8869c470af6e28bb1da027a54)

Conflicts:
	cinder/api/v3/volumes.py
	cinder/db/sqlalchemy/api.py
	cinder/exception.py
	cinder/volume/api.py
	cinder/volume/driver.py
	cinder/volume/manager.py
	cinder/volume/rpcapi.py
	etc/cinder/policy.json

xykong0926 committed:
Cinder volume revert to snapshot with Ceph

This commit implements reverting volume to a snapshot when
backend_storage is ceph

Change-Id: I845d4770f26ba6dc48bce2a4959b104df3668ea9
Partial-Implements: blueprint revert-volume-to-snapshot
(cherry picked from commit bf041fb7ffb38506402b49b8f3225c391c63fada)

zhaochao commented on cinder/api/contrib/volume_actions.py:
```diff
@@ -367,6 +367,41 @@ def _set_bootable(self, req, id, body):
+
+        context = req.environ['cinder.context']
+
+        params = body['os-revert']
```
这里要检查一下 body 是不是合法（不包含 os-revert ） ，上游实现其实也有，但已经封装接口，直接当前版本其它的 action 里类似的代码。

wwba replied:
没有直接掉is_valid_body（）有两个方面原因：
１，上游代码显示地raise HTTPBadRequest异常，＂params = body['os-revert']＂可能会抛出异常,后面＂snapshot_id = params.get("snapshot_id", None)＂才需要检查返回值；
２，参考_volume_upload_image()获取params;

此处改一下，捕获异常吧．

zhaochao replied:
没，我是说参考别的 action ，比如 os-retype ，直接取 body['os-revert']['snapshot_id'] ，如果出异常，或者取出来是空，都可以算是 BadRequest 。

zhaochao commented on cinder/api/contrib/volume_actions.py:
```diff
@@ -367,6 +367,41 @@ def _set_bootable(self, req, id, body):
+        msg = 'Reverting volume %(v_id)s to snapshot %(s_id)s.'
+        LOG.info(msg, {'v_id': id,
+                       's_id': snapshot_id})
+        self.volume_api.revert_to_snapshot(context, volume, snapshot)
```
需要保持上游代码里捕获 self.volume_api.revert_to_snapshot 拋出的异常的代码。

wwba replied:
上游代码之所以在这里处理InvalidVolume，InvalidSnapshot,VolumeSizeExceedsAvailableQuota,是重新进行了归类然后抛出，VolumeSizeExceedsAvailableQuota异常在这里不会抛出，HTTPConflict也没有必要进行Backport，所以这里捕获异常不是必须的，所有的异常都在api.py中捕获并进行处理即可．

zhaochao replied:
不是这样， cinder/volume/api.py 里面的异常不会直接和 HTTP 响应的异常对应，在 cinder/api 里要捕获异常，正是为了把后端的异常转换成前端的 HTTP 响应。

zhaochao commented on cinder/volume/api.py:
```diff
@@ -249,6 +249,40 @@ def create(self, context, size, name, description, snapshot=None,
+        
+        # Setting the status.
+        try:
+            self.db.snapshot_update(context, snapshot['id'],
```
更新数据库对象，可以不必特别捕获异常。

zhaochao commented on cinder/volume/api.py:
```diff
@@ -249,6 +249,40 @@ def create(self, context, size, name, description, snapshot=None,
+                   ) % msg_args
+            raise exception.BadResetResourceStatus(message=msg)
+　      try:
+            self.db.volume_update(context, volume['id'],
```
和上面相同，更新数据库对象，可以不必特别捕获异常。

zhaochao commented on cinder/volume/drivers/lvm.py:
```diff
@@ -252,6 +252,15 @@ def delete_snapshot(self, snapshot):
+        """Revert a volume to a snapshot"""
+
+        self.vg.revert(self._escape_snapshot(snapshot.name))
+        self.vg.deactivate_lv(volume.name)
```
deactivate_lv 是在 commit 314611f79afa6fe79f3d5022c814772048669bb0 中增加的，目前 cinder.brick.local_dev.lvm.LVM 没有这个方法。

zhaochao commented on cinder/volume/drivers/rbd.py:
```diff
@@ -684,6 +685,19 @@ def delete_volume(self, volume):
                 new_name = "%s.deleted" % (volume_name)
                 self.rbd.RBD().rename(client.ioctx, volume_name, new_name)
 
+    def revert_to_snapshot(self, context, volume, snapshot):
```
这个实现上游虽然没有合并，但是也是可以 cherry-pick 的，保留原始 commit 信息比较好。

如果保留原始 commit 信息，支持 revert 到任意 snapshot 是我们自己单独实现的，可以单独抽离出来作为第 3 个 commit 。

zhaochao commented on cinder/volume/manager.py:
```diff
@@ -514,6 +516,242 @@ def delete_volume(self, context, volume_id, unmanage_only=False):
+        self._notify_about_volume_usage(context, volume, "revert.end")
+        self._notify_about_snapshot_usage(context, snapshot, "revert.end")
+
+    def _create_snapshot_in_db(self, context,
```
这个函数大部份内容都是不需要的（状态校验这些在 api 应该都已经做完了），只保留创建 snapshot 的 options 以及 db.snapshot_create （db.snapshot_create 也不需要捕获异常）即可，最后和 _create_backup_snapshot 合并一个函数即可。这样就和上游的实现是一样的了。

wwba replied:
原来想封装成一个通用函数（必须进行volume和snapshot的状态检查），看来的确没有必要，这部分代码在_create_backup_snapshot中实现即可， 同时去掉配额管理．

zhaochao commented on cinder/volume/manager.py:
```diff
@@ -514,6 +516,242 @@ def delete_volume(self, context, volume_id, unmanage_only=False):
+
+        # if the backend driver do not support reverting to any snapshot,
+        # sure the snapshot is the latest one
+        support_reverting_to_any = False
```
1. 这个属性可以更明确一些，比如 support_reverting_to_any_snapshot ；
2. 这个属性为保持和上游实现的兼容，应该直接在 cinder.volume.driver.VolumeDriver 里增加为一个类属性，默认值为 False ，在 cinder.volume.drivers.rbd.RBDDriver 里设置为 True 就是对父类的覆盖。这样这里的代码只需要简单的判断是 self.support_reverting_to_any_snapshot 是不是 True 就可以了。

wwba replied:
你说的这个我之前想到了．
１，这样设计不随意修改上游cinder.volume.driver.VolumeDriver;
2, 新支持的support_reverting_to_any_snapshot driver才添加support_reverting_to_any_snapshot 属性．

zhaochao replied:
那这样倒是也可以，不过感觉没什么必要，毕竟和其它代码之间没有什么关联，直接改应该没什么影响。

zhaochao commented on cinder/volume/manager.py:
```diff
@@ -514,6 +516,242 @@ def delete_volume(self, context, volume_id, unmanage_only=False):
+        # Create a snapshot which can be used to restore the volume
+        # data by hand if revert process failed
+        try:
+            backup_snapshot = self._create_backup_snapshot(context, volume)
```
_creaet_backup_snapshot 和 _revert_to_snapshot 合到一个 try...except 里即可，和上游保持一致。分开的话，重复的代码比较多，因为 2 者任何一个出错，清理过程都是一样：

1. 发送 end 的 notification ；
2. 还原卷的状态；
~3. 删除临时快照。（这一步上游的代码好像是有点问题，可以加单独的 commit 修复）~
最后这一步不对，上游的实现是对的，临时快照本来是为了出问题时，手动恢复卷准备的，所以出现异常时，不能自动删除，只能人工检查后，再决定是否需要删除。

wwba replied:
特意把_creaet_backup_snapshot 和 _revert_to_snapshot 分开主要是方便调试的目的，creaet_backup_snapshot配额的问题是容易出错的． 
但是的确可以在一个放在try语句里面，根据backup_snapshot返回值判断哪里出现了异常．

zhaochao replied:
配额的问题后面说过了，这个过程中不要修改。由于 2 个步骤山错处理相同，还是应该合到一起。

zhaochao commented on cinder/volume/manager.py:
```diff
@@ -514,6 +516,242 @@ def delete_volume(self, context, volume_id, unmanage_only=False):
+        self._notify_about_snapshot_usage(context, snapshot_ref_now,
+                                          "revert.end")
+        try:
+            if volume['size'] != snapshot['size']:
```
这里的处理看起来有问题，按照 cinder/volume/driver.py 里 revert_to_snapshot 函数的注释，revert 是不改变 volume 的大小的，这种设计是合理的：

1. Cinder 里卷的大小只会扩大，不会缩小；
2. 如果卷的大小改变需要更新租户的 quota 使用量，增加复杂度；
3. 如果快照的大小比卷更大，为了防止数据丢失，应该拒绝 revert ；不过由于第 1 条的原因，不会出现快照比卷大的情况，所以不用考虑这种情况；
4. 如果快照的大小比卷更小，driver 在 revert 时把卷的大小变小（ rbd 即是这种情况），需要 rbd driver 里把卷的大小改回原来的大小（这一点上游待合并的 commit 没考虑到，我也去提了一个 comment ）

wwba replied:
这个２和４点我都没有考虑到,这里不更新volume size.

zhaochao commented on cinder/volume/manager.py:
```diff
@@ -514,6 +516,242 @@ def delete_volume(self, context, volume_id, unmanage_only=False):
+            msg = _("Revert finished, but failed to reset "
+                    "volume %(id)s status or size, "
+                    "please manually reset it.") % msg_args
+            raise exception.BadResetResourceStatus(message=msg)
```
噢，之前没有理解这个新增的 BadResetResourceStatus 的作用，这个是新版本在 update 数据库时增加的功能，即修改数据库会再次校验数据库对象的状态，这个功能现在版本没有，所以这个 Exception 看起来暂时没什么必要。

zhaochao commented on cinder/volume/manager.py:
```diff
@@ -514,6 +516,242 @@ def delete_volume(self, context, volume_id, unmanage_only=False):
+                    "please manually reset it.") % msg_args
+            raise exception.BadResetResourceStatus(message=msg)
+        try:
+            self.db.snapshot_update(context, snapshot['id'],
```
之前提过，更新数据库对象可以不对异常进行捕获。

zhaochao commented on cinder/volume/manager.py:
```diff
@@ -514,6 +516,242 @@ def delete_volume(self, context, volume_id, unmanage_only=False):
+            raise exception.BadResetResourceStatus(message=msg)
+
+        if backup_snapshot:
+            self.delete_snapshot(context, backup_snapshot['id'])
```
上游 commit 里，delete_snapshot 里关于 quota 的处理需要 backport ，即创建临时快照，不能占用 quota ，删除临时快照时相应的也不能对 quota 进行变动。

wwba committed:
Cinder volume support reverting to any snapshot

Cinder volume support reverting to any snapshot for some drivers.

Signed-off-by: Xiaojun Liao <xiaojunliao85@gmail.com>

wwba committed:
Resize rbd to match expected volume size when
reverting to a smaller snapshot

If the snapshot is created before the rbd extended, resize rbd.

Signed-off-by: Xiaojun Liao <xiaojunliao85@gmail.com>

zhaochao commented on cinder/brick/local_dev/lvm.py:
```diff
@@ -516,6 +516,21 @@ def _mangle_lv_name(self, name):
+            LOG.exception(_LE('Error deactivating LV'))
+            LOG.error(_LE('Cmd     :%s'), err.cmd)
+            LOG.error(_LE('StdOut  :%s'), err.stdout)
+            LOG.error(_LE('StdErr  :%s'), err.stderr)
```
`_LE` 这个没 import ，可以和这个文件其它地方保持一致只用 `_` 。

zhaochao commented on cinder/api/contrib/volume_actions.py:
```diff
@@ -367,6 +372,54 @@ def _set_bootable(self, req, id, body):
+
+        # Ensure snapshot is the volume's latest snapshot.
+        db_driver = CONF.db_driver
+        db = importutils.import_module(db_driver)
```
这里应该直接在文件开头 from cinder import db 就行。

zhaochao replied:
这里还没改，虽然这一部份后面的 commit 会去掉，但是还是要改下。

zhaochao commented on cinder/api/contrib/volume_actions.py:
```diff
@@ -367,6 +372,54 @@ def _set_bootable(self, req, id, body):
+        db_driver = CONF.db_driver
+        db = importutils.import_module(db_driver)
+        snapshot_db = db.snapshot_get_latest_for_volume(context,
+                                                        volume['id'])
```
db.snapshot_get_latest_for_volume 可能出现 VolumeSnapshotNotFound ，需要捕获。

zhaochao replied:
这里和上面 import db 一样，也需要改。变量名应该直接用上游 commit 里的 l_snap 。

zhaochao commented on cinder/exception.py:
```diff
@@ -869,3 +873,7 @@ class HBSDNotFound(NotFound):
     message = _("Bad response from Datera API")
+
+
+class BadResetResourceStatus(CinderException):
```
这个在当前的 Cinder 版本里用不上，可以去掉。

zhaochao commented on cinder/volume/manager.py:
```diff
@@ -514,6 +516,125 @@ def delete_volume(self, context, volume_id, unmanage_only=False):
+                   'encryption_key_id': volume['encryption_key_id'],
+                   'metadata': {}}
+
+        snapshot = None
```
上面这一行多余了。

zhaochao commented on cinder/volume/manager.py:
```diff
@@ -514,6 +516,125 @@ def delete_volume(self, context, volume_id, unmanage_only=False):
+        The process of reverting to snapshot consists of several steps:
+        1.   create a snapshot for backup (in case of data loss)
+        2.1. use driver's specific logic to revert volume
+        2.2. try the generic way to revert volume if driver's method is missing
```
generic way 相关的代码去掉了，这里注释也可以改了，最终在 commit msg 里也说明一下。

zhaochao commented on cinder/volume/manager.py:
```diff
@@ -514,6 +516,125 @@ def delete_volume(self, context, volume_id, unmanage_only=False):
+                   {'expected_status': 'reverting',
+                    'actual_status': actual_status})
+            self.db.volume_update(context, volume['id'],
+                                  {'status': 'available'})
```
这个地方还有点问题，应该是：
如果 volume 的状态不是 reverting ，中止操作，把 snapshot 的状态改为 available ，因为在 api 那里把 snapshot 改成 restoring 了。

这里改 volume 的状态是不对，因为已经发现 volume 的状态发生变化，不符合预期的 'reverting' 。

wwba replied:
这里不能改snapshot的状态，可以是除开restoring的另外一种状态，参考restore_backup， 这个是cinder-volume来做的，其实这里只需要将volume置为error状态，backup available即可（只读)

zhaochao commented on cinder/volume/manager.py:
```diff
@@ -514,6 +516,125 @@ def delete_volume(self, context, volume_id, unmanage_only=False):
+                   {'expected_status': 'restoring',
+                    'actual_status': actual_status})
+            self.db.snapshot_update(context, snapshot['id'],
+                                    {'status': 'available'})
```
同样这里也要改成：
如果 snapshot 的预期状态不对，应该中止操作，把 snapshot 的状态改成 error_restoring 之类的状态（这个状态的确是之前没有），同时把 volume 的状态改回 available （因为前面已经检查过 volume 的状态是 reverting ，说明是这个操作改动的，出错的话，得改回去）。

zhaochao commented on cinder/volume/manager.py:
```diff
@@ -514,6 +516,125 @@ def delete_volume(self, context, volume_id, unmanage_only=False):
+                                                "revert.end")
+                self._notify_about_snapshot_usage(context, snapshot_ref_now,
+                                                  "revert.end")
+                if backup_snapshot:
```
这里如果要在日志提醒 backup_snapshot 创建成功的话，可以放到这个异常处理的最后，在 msg 里加上，说明 backup_snapshot 已经创建成功。

不过其实意义不大。目前的实现，这个 backup_snapshot 创建成功之后只能手工想办法删除，没有可供调用的 api 来完成，最终是要人工手动处理。

wwba replied:
日志记录create_snapshot delete_snapshot执行完成后都会打印，这里的确不需要判断backup_snapshot的值．

zhaochao commented on cinder/volume/manager.py:
```diff
@@ -514,6 +516,125 @@ def delete_volume(self, context, volume_id, unmanage_only=False):
+                           'failed with %(error)s.')
+                    msg_args = {'v_id': volume['id'],
+                                'error': six.text_type(error)}
+                else:
```
这里就不是 else 了，因为上面出错了， revert 这个操作就失败了，下面这个 msg 是基本内容。

zhaochao commented on cinder/volume/manager.py:
```diff
@@ -514,6 +516,125 @@ def delete_volume(self, context, volume_id, unmanage_only=False):
+                                'error': six.text_type(error)}
+                LOG.exception(msg, msg_args)
+                self.db.volume_update(context, volume['id'],
+                                      {'status': 'error'})
```
这里也要同时处理 snapshot 的状态，改成 error_restoring 或者 available 。

更新：
这里 volume 的状态，改成 error_reverting 可能比较合适。

zhaochao commented on cinder/volume/manager.py:
```diff
@@ -590,6 +590,21 @@ def revert_to_snapshot(self, context, volume, snapshot):
+        # sure the snapshot is the latest one
+        if not (hasattr(self.driver, 'support_reverting_to_any_snapshot') and
+                self.driver.support_reverting_to_any_snapshot is True):
+            snapshot_db = self.db.snapshot_get_latest_for_volume(context,
```
snapshot_db 这种变量名不太好，可以用上游的 l_snap 。

zhaochao commented on cinder/volume/manager.py:
```diff
@@ -590,6 +590,21 @@ def revert_to_snapshot(self, context, volume, snapshot):
+                         'but got %(snap_id)s.') %
+                       {'snap_id': snapshot['id']})
+                self.db.snapshot_update(context, snapshot['id'],
+                                        {'status': 'available'})
```
这里也得还原 volume 的状态。

zhaochao commented on cinder/brick/local_dev/lvm.py:
```diff
@@ -516,6 +516,21 @@ def _mangle_lv_name(self, name):
+                          root_helper=self._root_helper,
+                          run_as_root=True)
+        except putils.ProcessExecutionError as err:
+            LOG.exception(_LE('Error deactivating LV'))
```
这里还留了一个 ``_LE``

zhaochao commented on cinder/volume/manager.py:
```diff
@@ -514,6 +516,140 @@ def delete_volume(self, context, volume_id, unmanage_only=False):
+                   {'expected_status': 'reverting',
+                    'actual_status': actual_status})
+            self.db.volume_update(context, volume['id'],
+                                  {'status': 'error_reverting'})
```
revert 操作失败之后，volume 和 snapshot 的状态设置，需要确定一个基本原则。

另外一个小问题是，每次调 volume_update 和 snapshot_update ，可以让它们的顺序保持一致。

zhaochao commented on cinder/volume/manager.py:
```diff
@@ -514,6 +516,140 @@ def delete_volume(self, context, volume_id, unmanage_only=False):
+                     "Try to use copy-snapshot-to-volume method.")
+            raise
+        self._notify_about_volume_usage(context, volume, "revert.end")
+        self._notify_about_snapshot_usage(context, snapshot, "revert.end")
```
~~revert.end 应该单独加一个 finally 这样的处理比较好。~~

看错了，后面有处理，这里不用了。
